src/memoization.ml

```diff
@@ -0,0 +1,376 @@
+}
+type cutoff_policy =
+  | No_cutoff
+  | Cutoff
+
+type 'a output_spec = {
+  serialize : 'a -> ser_output;
+  print : 'a -> string;
+  cutoff_policy : cutoff_policy;
+}
+
+(* a function that will try to recompute the memoized function to ensure
+   the output is up-to-date *)
+type update_cache = unit -> ser_output Fiber.t
+
+module RunId = Id.Make ()
+module FnId = Id.Make ()
+module Id = Id.Make()
+
+module Kind = struct
+  type 'a tag = ..
+
+  module type S = sig
+    type t
+    type 'a tag += X : t tag
+  end
+
+  type 'a t = (module S with type t = 'a)
+
+  let create (type a) () =
+    let module M = struct
+      type t = a
+      type 'a tag += X : t tag
+    end in
+    (module M : S with type t = a)
+
+  let eq (type a) (type b)
+        (module A : S with type t = a)
+        (module B : S with type t = b)
+    : (a, b) Type_eq.t option =
+    match A.X with
+    | B.X -> Some Type_eq.T
+    | _   -> None
+end
+
+module Output = struct
+  type 'output run_state =
+    | Running of 'output Fiber.Ivar.t
+    | Done of 'output
+
+  type 'a output_cache = {
+    state : 'a run_state;
+    kind : 'a Kind.t;
+  }
+
+  type packed_output = Packed : _ output_cache -> packed_output
+
+  let unpack : type a. a Kind.t -> packed_output -> a run_state =
+    fun kind packed ->
+      match packed with
+      |  Packed output ->
+        let Type_eq.T = Kind.eq (output.kind) kind |> Option.value_exn in
+        output.state
+
+  let pack kind v = Packed {
+    state = v;
+    kind = kind;
+  }
+end
+
+type global_cache_info = {
+  last_output : ser_output;
+  mutable last_run : RunId.t;
+  update : update_cache;
+  cutoff_policy : cutoff_policy;
+  last_deps : (dep_node * ser_output) list;
+}
+
+and dep_info = {
+  name : name;
+  input : ser_input;
+  id : Id.t;
+  mutable cache : global_cache_info option;
+  mutable output : Output.packed_output option;
+}
+
+and dep_node = dep_info Dag.node
+
+module Stack_frame = struct
+  type t = {
+    dep_node : dep_node;
+  }
+
+  let di v = Dag.get v.dep_node
+
+  let name v = (di v).name
+  let input v = (di v).input
+end
+
+module Cycle_error = struct
+  type t = {
+    cycle : dep_node list;
+    stack : Stack_frame.t list;
+  }
+
+  exception E of t
+
+  let stack ex = ex.stack
+
+  let filter ~(name:name) ex =
+    ex.cycle
+    |> List.filter ~f:(fun (f : dep_node) -> (Dag.get f).name = name)
+    |> List.map ~f:(fun (f : dep_node) -> (Dag.get f).input)
+
+  let serialize ex =
+    ex.cycle
+    |> List.map ~f:Dag.get
+    |> List.map ~f:(fun (f : dep_info) -> Format.sprintf "%s %s" f.name f.input)
+
+  let serialize_stack ex =
+    ex.stack
+    |> List.map ~f:(fun v -> Stack_frame.di v)
+    |> List.map ~f:(fun (f : dep_info) -> Format.sprintf "%s %s" f.name f.input)
+end
+
+
+let global_dep_dag = Dag.create ()
+let global_dep_table : (FnId.t * ser_input, dep_node) Hashtbl.t = Hashtbl.create 256
+
+let global_id_table : (name, FnId.t) Hashtbl.t = Hashtbl.create 256
```

Comment:
Instead of FnId, you can use something like `module Function_name = Interned.Make(...)`

Reply:
Changed.